### PR TITLE
feat: use quote endpoint for THORSwap data

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Updates the THORSwapper trade quote logic to use data from the quote endpoint instead of hand-rolling calculations, including those relating to:

- slippage
- buy amount
- fees

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Partially closes https://github.com/shapeshift/web/issues/4360 (we still need to replace our limit/memo logic, but that's for another PR)

## Risk

Reasonably high with respect to THORSwapper trades.

## Testing

- Check THORSwapper quote/trade values look as expected, ensuring RUNE and non-RUNE trades are checked
- Check that THORSwapper trades complete as expected

### Engineering

☝️

In the response to the `lcd/thorchain/quote/swap` request, check that `expected_amount_out`  is the roughly same as the "you get" amount. It won't quite match as we store fees in USD, and so we are converting to and from before the state layer, and so we lose some precision.

The "amount before fees" should be the "you get" amount (factoring slippage) + `outbound` and `affiliate` fee amounts in the `fees` object.

### Operations

☝️

## Screenshots (if applicable)

N/A

### CODE REVIEW

**Comment by 0xApotheosis** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4361** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4361" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4361?utm_source=stack-comment).

File: src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.test.ts [hunk 1]
```diff
@@ -65,12 +65,12 @@ describe('getTradeQuote', () => {
           return Promise.resolve({
             data: {
               dust_threshold: '10000',
-              expected_amount_out: '261454522054192',
+              expected_amount_out: '10232161',
               expiry: 1681132269,
               fees: {
                 affiliate: '0',
-                asset: 'ETH.FOX-0XC770EEFAD204B5180DF6A14EE197D99D808EE52D',
```
Comment (0xApotheosis) [inline src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.test.ts:72]: Our mocked quote response here was for an ETH to FOX trade, but the tests are expecting a FOX to ETH trade. As we are now using this data to compute the quote details, this had to be updated (and expected test values changed accordingly).

File: src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts [hunk 2]
```diff
@@ -100,48 +117,54 @@ export const getThorTradeQuote: GetThorTradeQuote = async ({ deps, input }) => {
...
-
-    const recommendedSlippage = await getSlippage({
-      inputAmountThorPrecision: bn(sellAmountThorPrecision),
-      daemonUrl: deps.daemonUrl,
-      buyAssetId: buyAsset.assetId,
-      sellAssetId: sellAsset.assetId,
-    })
+    const buyAssetTradeFeeBuyAssetCryptoHuman = fees
+      ? fromBaseUnit(bnOrZero(fees.outbound), THORCHAIN_FIXED_PRECISION)
+      : await getInboundAddressDataForChain(deps.daemonUrl, buyAsset.assetId).then(data =>
```
Comment (0xApotheosis) [inline src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts:124]: Only do this network request if we need to.

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    return Err(
+      makeSwapErrorRight({
+        message: `[getTradeRate]: THORChain quote returned an error: ${data.error}`,
+        code: SwapErrorType.TRADE_QUOTE_FAILED,
+        details: { sellAssetId: sellAsset.assetId, buyAssetId },
+      }),
+    )
+  } else {
+    return Ok(data)
+  }
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts:70]: if you prefer guard clause style

```suggestion
  }

  if ('error' in data) {
    return Err(
      makeSwapErrorRight({
        message: `[getTradeRate]: THORChain quote returned an error: ${data.error}`,
        code: SwapErrorType.TRADE_QUOTE_FAILED,
        details: { sellAssetId: sellAsset.assetId, buyAssetId },
      }),
    )
  }

  return Ok(data)
```

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  const sellAmountCryptoThorBaseUnit = bn(
+    toBaseUnit(sellAmountCryptoPrecision, THORCHAIN_FIXED_PRECISION),
+  )
+
+  const { data } = await thorService.get<ThornodeQuoteResponse>(
+    `${deps.daemonUrl}/lcd/thorchain/quote/swap?amount=${sellAmountCryptoThorBaseUnit}&from_asset=${sellPoolId}&to_asset=${buyPoolId}&destination=${receiveAddress}&affiliate_bps=${affiliateBps}&affiliate=ss`,
+  )
+
+  if ('error' in data && /not enough fee/.test(data.error)) {
+    // TODO(gomes): How much do we want to bubble the error property up?
```
Comment (0xApotheosis) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts:50]: The bulk of this is the same logic as already exists in `getTradeRate.ts`, though this helper returns the entire response.

File: src/lib/swapper/swappers/ThorchainSwapper/utils/txFeeHelpers/utxoTxFees/getUtxoTxFees.ts [hunk 5]
```diff
@@ -12,6 +12,7 @@ type GetUtxoTxFeesInput = {
   sellAdapter: UtxoBaseAdapter<ThorUtxoSupportedChainId>
   pubkey: string
   buyAssetTradeFeeUsd: string
+  sellAssetTradeFeeUsd: string
```
Comment (0xApotheosis) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/txFeeHelpers/utxoTxFees/getUtxoTxFees.ts:15]: Plumbing for affiliate fees.

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  const sellAmountCryptoPrecision = bn(sellAmountCryptoBaseUnit).div(
+    bn(10).pow(sellAsset.precision),
+  )
+  // All THORChain pool amounts are base 8 regardless of token precision
+  const sellAmountCryptoThorBaseUnit = bn(
+    toBaseUnit(sellAmountCryptoPrecision, THORCHAIN_FIXED_PRECISION),
+  )
+
+  const { data } = await thorService.get<ThornodeQuoteResponse>(
+    `${deps.daemonUrl}/lcd/thorchain/quote/swap?amount=${sellAmountCryptoThorBaseUnit}&from_asset=${sellPoolId}&to_asset=${buyPoolId}&destination=${receiveAddress}&affiliate_bps=${affiliateBps}&affiliate=ss`,
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts:46]: if `qs` preferable for encoding query strings?

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  sellAmountCryptoBaseUnit: string
+  receiveAddress: string
+  affiliateBps?: string
+  deps: ThorchainSwapperDeps
+}): Promise<Result<ThornodeQuoteResponseSuccess, SwapErrorRight>> => {
+  const buyPoolId = adapters.assetIdToPoolAssetId({ assetId: buyAssetId })
+  const sellPoolId = adapters.assetIdToPoolAssetId({ assetId: sellAsset.assetId })
+
+  const sellAmountCryptoPrecision = bn(sellAmountCryptoBaseUnit).div(
+    bn(10).pow(sellAsset.precision),
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts:38]: we can use `baseUnitToHuman` if you prefer

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts:70]: Heh, I deliberately removed the guard clause pattern. Maybe it's my functional background, but I like to have only one return point in functions where it makes sense to do so (e.g. here, where all returns fit nicely in an `if...else`.

File: src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts [hunk 8]
```diff
@@ -79,16 +74,38 @@ export const getThorTradeQuote: GetThorTradeQuote = async ({ deps, input }) => {
...
+    const rate = await quote.match({
+      ok: quote => {
+        const THOR_PRECISION = 8
+        const expectedAmountOutThorBaseUnit = quote.expected_amount_out
+        // Add back the outbound fees
+        const expectedAmountPlusFeesCryptoThorBaseUnit = bn(expectedAmountOutThorBaseUnit).plus(
+          quote.fees.outbound,
+        )
+        const sellAmountCryptoPrecision = bn(sellAmountCryptoBaseUnit).div(
+          bn(10).pow(sellAsset.precision),
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts:94]: we can use `baseUnitToPrecision` if you prefer

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts (hunk 3 above)
Comment (woodenfurniture) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts:70]: ah right :P im on the other end of the spectrum where i can "stop reading" after a certain point

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+      }),
+    )
+  } else if ('error' in data) {
+    return Err(
+      makeSwapErrorRight({
+        message: `[getTradeRate]: THORChain quote returned an error: ${data.error}`,
+        code: SwapErrorType.TRADE_QUOTE_FAILED,
+        details: { sellAssetId: sellAsset.assetId, buyAssetId },
+      }),
+    )
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts:67]: (side note, not blocking) i could picture an future state where we ditched `SwapError` and the `SwapError.code` altogether and simply bubbled up error messages to display in the UI. In this case we'd be able to remove the multiple cases and simply map the thorchain error message to a UI friendly string to return

File: src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts (hunk 8 above)
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts]: Addressed in https://github.com/shapeshift/web/pull/4361/commits/14a26c4702a6da0f999cdb4a3987ad25bf7091b2.

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts (hunk 7 above)
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts]: Addressed in https://github.com/shapeshift/web/pull/4361/commits/14a26c4702a6da0f999cdb4a3987ad25bf7091b2.

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts (hunk 6 above)
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts]: I like it - used `qs` in https://github.com/shapeshift/web/pull/4361/commits/fdd6eaadcb83f2708e845d022d0eedf8eb66ef2c.

File: src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts (hunk 9 above)
Comment (0xApotheosis) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/getQuote/getQuote.ts:80]: Yeah indeed, we need to think through this one a little more.

**Review by woodenfurniture — APPROVED** [review]
code reviewed, feedback actioned, fees looking perfect

**Pull request merged**